torch/csrc/jit/script/compiler.cpp

```diff
@@ -899,12 +935,30 @@ struct to_ir {
     // Register outputs in each block
     for (const auto& x : mutated_variables) {
       auto tv = save_true->getVar(x, stmt.range());
-      true_block->registerOutput(tv);
       auto fv = save_false->getVar(x, stmt.range());
+      auto unified = unifyTypes(tv->type(), fv->type());
+
+      // attempt to unify the types. we allow variables to be set to different types
+      // in each branch as long as that variable is not already in scope,
+      // or if that variable does not get used later. here, we save the error
+      // so that the error message will be more informative in the case that is
+      // used later
+
+      if (!unified) {
+        ErrorReport error(stmt);
+        error << "Type mismatch: " << x << " is set to type " << tv->type()->str() << " in the true branch"
+        << " and type " << fv->type()->str() << " in the false branch";
+        if (save_true->findInParentFrame(x) || save_false->findInParentFrame(x)) {
+          throw error;
+        } else {
+          save_true->setVariableTypeError(x, error.what());
```

Comment:
ah i'll comment that. Since it's saved in the lowest level environment (bc all variables are in scope), it doesn't matter whether that gets accessed through save_true or save_false.